File: src/09.tree/50.110.go
package tree

import "math"

func IsBalanced(root *TreeNode) bool {
	var dfs func(node *TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}

		l := dfs(node.Left)
		if l == -1 {
			return -1
		}

		r := dfs(node.Right)
		if r == -1 {
			return -1
		}

		if math.Abs(float64(l-r)) > 1 {
			return -1
		}

		if l > r {
			return l + 1
		}

		return r + 1
	}
	return dfs(root) != -1
}

File: src/09.tree/50.110.ts
import { TreeNode } from "./01.100";

export function isBalanced(root: TreeNode | null): boolean {
    const dfs = (node: TreeNode | null) => {
        if (!node) {
            return 0;
        }

        const l = dfs(node.left);
        if (l === -1) {
            return -1;
        }

        const r = dfs(node.right);
        if (r === -1) {
            return -1;
        }

        return Math.abs(l - r) > 1 ? -1 : Math.max(l, r) + 1;
    };

    return dfs(root) != -1;
}
